README.md
# cloud-search-query
An ORM-like wrapper for building AWS CloudSearch structured queries

## Installation ##
CloudSearchQuery is currently a repository package only.  In `composer.json` add:

    "repositories": [
      {
        "type": "vcs",
        "url": "https://github.com/aaronkaz/cloud-search-query.git"
      }
    ],
    "require": {
        "aaron-kaz/cloud-search-query": "dev-master"
    },


## Basic Usage ##
Initialize a query object with a valid CloudSearch full URI endpoint

    $query = new CloudSearchQuery([
        'endpoint' => 'http://search-yourdomain.us-east-1.cloudsearch.amazonaws.com'
    ]);

You can chain query methods like so

    $query->phrase('ford')
          ->term('National Equipment', 'seller')
          ->range('year', '1987');

use the `get()` method to submit query and retrieve results from AWS.  Use property accessors on the returned results object.

    $results = $query->get();
    $matchedDocuments = $results->hits;

## Search Query Operators and Nested Queries ##
You can use the `and`, `or`, and `not` operators to build compound and nested queries.  
The corresponding `and()`, `or()`, and `not()` methods expect a closure as their argument.
You can chain all available methods as well nest more subqueries inside of closures.

    $query->or(function($builder) {
              $builder->phrase('ford')
                      ->phrase('truck');
            })


CHANGELOG.md
## [2.2.2] - 2017-05-29
### Changed
- Pass config as to CloudSearchQuery constructor.  @aaronkaz
- Pass CloudSearchQuery instance as parent to Builder.  @aaronkaz


## [2.2.0] - 2017-03-08
### Added
- Add cursor building for queries exceeding 10000 offset limit.  @aaronkaz


## [2.1.2] - 2017-02-27
### Fixed
- Ensure \ characters are escaped before other chars.  @aaronkaz


## [2.1.1] - 2017-01-19
### Fixed
- Escape \ character in search strings.  @aaronkaz


tests/TestCase.php
<?php

namespace TheLHC\CloudSearchQuery\Tests;

use Orchestra\Testbench\TestCase as BaseTestCase;
use TheLHC\CloudSearchQuery\CloudSearchQueryServiceProvider;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Database\Schema\Builder;
use Dotenv\Dotenv;

class TestCase extends BaseTestCase
{

    /**
     * Setup the test environment.
     *
     * @return void
     */
    public function setUp()
    {
        parent::setUp();
    }

    /**
     * Define environment setup.
     *
     * @param  \Illuminate\Foundation\Application $app
     * @return void
     */
    protected function getEnvironmentSetUp($app)
    {
        $dotenv = new Dotenv(__DIR__);
        $dotenv->load();
        $app['config']->set(
            'cloud_search_query.endpoint',
            getenv('ENDPOINT')
        );
    }

    /**
     * Get package service providers.
     *
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [
            CloudSearchQueryServiceProvider::class
        ];
    }

}


tests/CloudSearchQueryTest.php
<?php

use TheLHC\CloudSearchQuery\CloudSearchQuery;
use TheLHC\CloudSearchQuery\Tests\TestCase;

class CloudSearchQueryTest extends TestCase
{

    public function testCanResolveFromTheContainer()
    {
        $manager = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $this->assertInstanceOf(CloudSearchQuery::class, $manager);
    }

    public function testItSearchesWithBreakingChars()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->term("Vander Haag's", 'seller');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
    }

    public function testItSearchWithQuotesInParams()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->term("TSE INTERNATIONAL", 'make')
                ->term('82" HEAVY DUTY TILLER');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
    }

    public function testItSearchesWithNearSearch()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->near("2012 FORD F-150");
        $results = $query->get();
        $this->assertEquals('200', $results->status);
    }

    public function testItSearchesWithOptions()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->options('fields', ['seller', 'category']);
        $query->qOr(function($builder) {
            $term = "Gregory Poole Equipment";
            $builder->phrase("{$term}*");
            $builder->prefix($term);
            $builder->near($term);
        });
        $query->facet('seller');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
        $this->assertEquals(false, empty($results->hits));
    }

    public function testItSearchesWithNestedPhrase()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->size(10)
              ->start(0)
              ->qOr(function($builder) {
                $builder->phrase('ford')
                        ->phrase('truck');
              })
              ->term('National Equipment', 'seller')
              ->range('year', '1987');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
    }

    public function testItSearchesWithFilterQuery()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->size(10)
              ->start(0)
              ->filterOr(function($builder) {
                $builder->phrase('ford', 'desc')
                        ->phrase('truck', 'desc');
              })
              ->filterTerm('National Equipment', 'seller')
              ->range('year', '1987');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
    }

    public function testItSearchesWithLatLon()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->phrase('ford')->latlon('latlon', '34.707731', '-89.906631');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
    }

    public function testItSearchesWithFacets()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->phrase('ford')->facet('make', 'count')->facet('model_family');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
        $this->assertEquals(true, is_array($results->facets));
    }

    public function testItSearchesWithExpression()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $lat = '34.707731';
        $lon = '-89.906631';
        $query->phrase('ford')
              ->latlon('latlon', $lat, $lon, 50, true)
              ->sort('distance', 'asc');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
    }

    public function testItSearchesWithStats()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->phrase('ford')
                ->facet('year', 'count')
                ->stats('year');
        $results = $query->get();
        $this->assertEquals('200', $results->status);
        $this->assertEquals(true, is_array($results->stats));

    }

    public function testItSearchesWithFacetBuckets()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->phrase('ford')->facetBuckets('year', ["[1970,1979]","[1980,1989]","[1990,1999]"]);
        $results = $query->get();
        $this->assertEquals('200', $results->status);
        $this->assertEquals(true, is_array($results->facets));
    }

    public function testItCatchesMaxStart()
    {
        $query = $this->app->make('TheLHC\CloudSearchQuery\CloudSearchQuery');
        $query->size(10)
              ->start(15000)
              ->matchall()
              ->pretty();
        $results = $query->get();
        $this->assertEquals('200', $results->status);
        $this->assertEquals(true, is_array($results->hits));
    }

}


src/StructuredSearch.php
<?php

namespace TheLHC\CloudSearchQuery;

class StructuredSearch
{

    /**
     * Array of structured search statements
     *
     * @var Array
     */
    public $query;

    /**
     * Overloading method
     * Query and filterQuery methods are made inaccessible in order to trigger
     * the __call magic method.  Here we can escape argument strings prior to
     * calling the intended function
     *
     * @param  string $method
     * @param  array $args
     * @return StructuredQueryBuilder
     */
    public function __call($method, $args) {
        // raise exception if no method
        if (!method_exists($this, $method)) {
            throw new Exception("Method doesn't exist");
        }
        // escape string arguments
        foreach($args as $key => $value) {
            if (gettype($value) == "string") {
                $cleanValue = $value;
                $cleanValue = preg_replace('#\\\#i', "\\\\\\", $cleanValue);
                $cleanValue = preg_replace("#'#", "\\'", $cleanValue);
                $args[$key] = $cleanValue;
            }
        }
        // call the desired the method
        call_user_func_array([$this, $method], $args);
        return $this;
    }

    /**
     * Magic method when casting as string
     * Concatenate all query statements into wrapped 'and' statement
     *
     * @return string
     */
    public function __toString()
    {
        return "(and ".implode('', $this->query).")";
    }

    /**
     * Get the query property
     *
     * @return Array
     */
    public function getQuery()
    {
        return $this->query;
    }

    /**
     * AND OPERATOR
     * Includes a document only if it matches all of the specified expressions.
     * (Boolean AND operator.) The expressions can contain any of the structured
     * query operators, or a simple search string.
     *
     * @param  function|string $block
     * @return StructuredSearch
     */
    public function qAnd($block)
    {
        if (gettype($block) == "object") {
            $builder = new $this;
            $block($builder);
            $and = "(and ".implode('', $builder->getQuery()).")";
        } elseif (gettype($block) == "string") {
            $and = "(and '{$block}')";
        }
        $this->query[] = $and;
        return $this;
    }

    /**
     * MATCHALL
     * Matches every document in the domain.
     *
     * @return StructuredSearch
     */
    public function matchall()
    {
        $this->query[] = "(matchall)";
    }

    /**
     * NEAR OPERATOR
     * Searches a text or text-array field for the specified multi-term string and matches documents that contain the terms within the specified distance of one another. (This is sometimes called a sloppy phrase search.) If you omit the field option, Amazon CloudSearch searches all statically configured text and text-array fields by default. Dynamic fields and literal fields are not searched by default. You can specify which fields you want to search by default by specifying the q.options fields option.
     *
     * @param  string  $value
     * @param  string  $field
     * @param  integer $distance
     * @param  integer  $boost
     * @return StructuredSearch
     */
    private function near($value, $field = null, $distance = 3, $boost = null)
    {
        $near = "(near ";
        if ($field) {
            $near .= "field='{$field}' ";
        }
        if ($distance) {
            $near .= "distance='{$distance}' ";
        }
        if ($boost) {
            $near .= "boost='{$boost}' ";
        }
        $near .= "'{$value}')";
        $this->query[] = $near;
        return $this;
    }

    /**
     * NOT OPERATOR
     * Excludes a document if it matches the specified expression. (Boolean NOT
     * operator.) The expression can contain any of the structured query
     * operators, or a simple search string.
     *
     * @param  function|string $block
     * @return StructuredSearch
     */
    public function qNot($block)
    {
        if (gettype($block) == "object") {
            $builder = new $this;
            $block($builder);
            $not = "(not ".implode('', $builder->getQuery()).")";
        } elseif (gettype($block) == "string") {
            $not = "(not '{$block}')";
        }
        $this->query[] = $not;
        return $this;
    }

    /**
     * OR OPERATOR
     * Includes a document if it matches any of the specified expressions.
     * (Boolean OR operator.) The expressions can contain any of the structured
     * query operators, or a simple search string.
     *
     * @param  function|string $block
     * @return StructuredSearch
     */
    public function qOr($block)
    {
        if (gettype($block) == "object") {
            $builder = new $this;
            $block($builder);
            $or = "(or ".implode('', $builder->getQuery()).")";
        } elseif (gettype($block) == "string") {
            $or = "(or '{$block}')";
        }
        $this->query[] = $or;
        return $this;
    }

    /**
     * PHRASE OPERATOR
     * Searches a text or text-array field for the specified phrase. If you
     * omit the field option, Amazon CloudSearch searches all statically
     * configured text and text-array fields by default. Dynamic fields and
     * literal fields are not searched by default. You can specify which fields
     * you want to search by default by specifying the q.options fields option.
     *
     * @param  string $value
     * @param  string $field
     * @param  integer $boost
     * @return StructuredSearch
     */
    private function phrase($value, $field = null, $boost = null)
    {
        $phrase = "(phrase ";
        if ($field) {
            $phrase .= "field='{$field}' ";
        }
        if ($boost) {
            $phrase .= "boost='{$boost}' ";
        }
        $phrase .= "'{$value}')";
        $this->query[] = $phrase;
        return $this;
    }

    /**
     * PREFIX OPERATOR
     * Searches a text, text-array, literal, or literal-array field for the
     * specified prefix followed by zero or more characters. If you omit the
     * field option, Amazon CloudSearch searches all statically configured text
     * and text-array fields by default. Dynamic fields and literal fields are
     * not searched by default. You can specify which fields you want to search
     * by default by specifying the q.options fields option.
     *
     * @param  string $value
     * @param  string $field
     * @param  integer $boost
     * @return StructuredSearch
     */
    private function prefix($value, $field = null, $boost = null)
    {
        $prefix = "(prefix ";
        if ($field) {
            $prefix .= "field='{$field}' ";
        }
        if ($boost) {
            $prefix .= "boost='{$boost}' ";
        }
        $prefix .= "'{$value}')";
        $this->query[] = $prefix;
        return $this;
    }

    /**
     * RANGE OPERATOR
     * Searches a numeric field (double, double-array, int, int-array) or date
     * field (date, date-array) for values in the specified range. Matches
     * documents that have at least one value in the field within the specified
     * range. The field option must be specified.
     *
     * To specify a range of values, use a comma (,) to separate the upper and
     * lower bounds and enclose the range using brackets or braces. A square
     * bracket, [ or ], indicates that the bound is included in the range, a
     * curly brace, { or }, excludes the bound. You can omit the upper or lower
     * bound to specify an open-ended range. When omitting a bound, you must
     * use a curly brace.
     *
     * Dates and times are specified in UTC (Coordinated Universal Time)
     * according to IETF RFC3339: yyyy-mm-ddTHH:mm:ss.SSSZ. In UTC, for example,
     * 5:00 PM August 23, 1970 is: 1970-08-23T17:00:00Z. Note that you can also
     * specify fractional seconds when specifying times in UTC. For example,
     * 1967-01-31T23:20:50.650Z.
     *
     * @param  string $field
     * @param  integer|string $min
     * @param  integer|string $max
     * @return StructuredSearch
     */
    public function range($field, $min, $max)
    {
        $range = "(range field={$field} ";
        if ($min and !$max) {
            $value = "[{$min},}";
        } elseif (!$min and $max) {
            $value = "{,{$max}]";
        } elseif ($min and $max) {
            $value = "[{$min},{$max}]";
        } else {
            return;
        }
        $range .= "{$value})";
        $this->query[] = $range;
        return $this;
    }

    /**
     * TERM OPERATOR
     * Searches the specified field for a string, numeric value, or date. The
     * field option must be specified when searching for a value. If you omit
     * the field option, Amazon CloudSearch searches all statically configured
     * text and text-array fields by default. Dynamic fields and literal fields
     * are not searched by default. You can specify which fields you want to
     * search by default by specifying the q.options fields option.
     *
     * Dates and times are specified in UTC (Coordinated Universal Time)
     * according to IETF RFC3339: yyyy-mm-ddTHH:mm:ss.SSSZ. In UTC, for example,
     * 5:00 PM August 23, 1970 is: 1970-08-23T17:00:00Z. Note that you can also
     * specify fractional seconds when specifying times in UTC. For example,
     * 1967-01-31T23:20:50.650Z.
     *
     * @param  string $value
     * @param  string $field
     * @param  integer $boost
     * @return StructuredSearch
     */
    private function term($value, $field = null, $boost = null)
    {
        $term = "(term ";
        if ($field) {
            $term .= "field='{$field}' ";
        }
        if ($boost) {
            $term .= "boost='{$boost}' ";
        }
        $term .= "'{$value}')";
        $this->query[] = $term;
        return $this;
    }


}


src/StructuredQueryBuilder.php
<?php

namespace TheLHC\CloudSearchQuery;

use Aws\CloudSearchDomain\CloudSearchDomainClient;
use Aws\CloudSearchDomain\Exception\CloudSearchDomainException;
use Cache;

class StructuredQueryBuilder {

    /**
     * Parent CloudSearchQuery if passed
     *
     * @var CloudSearchQuery
     */
    private $parent;

    /**
     * Cursor value
     *
     * @var string
     */
    public $cursor;

    /**
     * Array of key-value pairs for custom expressions
     *
     * @var Array
     */
    public $expressions = [];

    /**
     * Array of facet specifications
     *
     * @var Array
     */
    public $facets = [];

    /**
     * Content type of response
     *
     * @var string
     */
    public $format = 'json';

    /**
     * Specifies a structured query that filters the results of a search without
     * affecting how the results are scored and sorted. You use fq in
     * conjunction with the q parameter to filter the documents that match the
     * constraints specified in the q parameter. Specifying a filter just
     * controls which matching documents are included in the results, it has no
     * effect on how they are scored and sorted.
     *
     * @var StructuredSearch
     */
    public $fq;

    // FUTURE METHODS
    //
    //public $highlights;
    //
    //public $partial;

    /**
     * Format json so it's easier to read
     *
     * @var boolean
     */
    public $pretty;

    /**
     * The search criteria for the request
     *
     * @var StructuredSearch
     */
    public $q;

    /**
     * Query parser options
     *
     * @var Array|JSON object
     */
    public $options;

    /**
     * Fields to return in results
     *
     * @var string
     */
    public $returnFields;

    /**
     * Number results to return per page
     *
     * @var integer
     */
    public $size = 10; // default per page

    /**
     * Fields/expressions to sort by
     *
     * @var string
     */
    public $sort;

    /**
     * Search results offset
     *
     * @var integer
     */
    public $start = 0; // default offset

    /**
     * Array of fields to get statistics
     *
     * @var Array
     */
    public $stats;


    /**
     * Create new instance
     */
    public function __construct($parent = null)
    {
        if ($parent) $this->parent = $parent;
        $this->q = new StructuredSearch();
        $this->fq = new StructuredSearch();
    }

    /**
     * Alias to get structured search query
     *
     * @return array
     */
    public function getQuery()
    {
        return $this->q->getQuery();
    }

    /**
     * Alias to get structured filter query
     *
     * @return array
     */
    public function getFilterQuery()
    {
        return $this->fq->getQuery();
    }

    /**
     * CURSOR
     * Retrieves a cursor value you can use to page through large result sets.
     * Use the size parameter to control the number of hits you want to include
     * in each response. You can specify either the cursor or start parameter in
     * a request, they are mutually exclusive.
     *
     * To get the first cursor, specify cursor=initial in your initial request.
     * In subsequent requests, specify the cursor value returned in the hits
     * section of the response.
     *
     * @param  string $cursor
     * @return StructuredQueryBuilder
     */
    public function cursor($cursor = 'initial')
    {
        $this->cursor = $cursor;
        return $this;
    }

    /**
     * EXPRESSION
     * Defines an expression that can be used to sort results. You can also
     * specify an expression as a return field.
     *
     * @param  string $accessor
     * @param  string $expression
     * @return StructuredQueryBuilder
     */
    public function expr($accessor, $expression)
    {
        $this->expressions[$accessor] = $expression;
    }

    /**
     * FACET (sorted)
     * Specifies a field that you want to get facet information for—FIELD is the
     * name of the field. The specified field must be facet enabled in the
     * domain configuration. Facet options are specified as a JSON object. If
     * the JSON object is empty, facet.FIELD={}, facet counts are computed for
     * all field values, the facets are sorted by facet count, and the top 10
     * facets are returned in the results.
     *
     * sort specifies how you want to sort the facets in the results: bucket or
     * count. Specify bucket to sort alphabetically or numerically by facet
     * value (in ascending order). Specify count to sort by the facet counts
     * computed for each facet value (in descending order).
     *
     * size specifies the maximum number of facets to include in the results.
     * By default, Amazon CloudSearch returns counts for the top 10.
     *
     * @param  string  $field
     * @param  string  $sort
     * @param  integer $size
     * @return StructuredQueryBuilder
     */
    public function facet($field, $sort = "bucket", $size = 10)
    {
        $this->facets[$field] = [
            'sort' => $sort,
            'size' => $size
        ];
    }

    /**
     * FACET (BUCKETS)
     * specifies an array of the facet values or ranges you want to count.
     * Buckets are returned in the order they are specified in the request. To
     * specify a range of values, use a comma (,) to separate the upper and
     * lower bounds and enclose the range using brackets or braces. A square
     * bracket, [ or ], indicates that the bound is included in the range, a
     * curly brace, { or }, excludes the bound. You can omit the upper or lower
     * bound to specify an open-ended range. When omitting a bound, you must
     * use a curly brace.
     *
     * @param  string $field
     * @param  Array $buckets
     * @param  string $method
     * @return StructuredQueryBuilder
     */
    public function facetBuckets($field, $buckets, $method = "filter")
    {
        $this->facets[$field] = [
            'buckets' => $buckets,
            'method'  => $method
        ];
    }

    /**
     * PRETTY
     * Formats JSON output so it's easier to read.
     *
     * @return StructuredQueryBuilder
     */
    public function pretty()
    {
        $this->pretty = true;
        return $this;
    }

    /**
     * QUERY PARSER OPTIONS
     * Configure options for the query parser specified in the q.parser
     * parameter.	The options are specified as a JSON object, for example:
     * q.options={defaultOperator: 'or', fields: ['title^5','description']}
     *
     * defaultOperator-The default operator used to combine individual terms
     * in the search string. (and|or)
     * defaultOperator: 'or'
     *
     * fields—An array of the fields to search when no fields are specified in
     * a search.  You can specify a weight for each field to control the relative
     * importance of each field when Amazon CloudSearch calculates relevance scores.
     * fields: ['title^5','description']
     *
     * @param  string $key
     * @param  mixed $value
     * @return StructuredQueryBuilder
     */
    public function options($key, $value)
    {
        $this->options[$key] = $value;
        return $this;
    }

    /**
     * RETURN FIELDS
     * The field and expression values to include in the response, specified as
     * a comma-separated list. By default, a search response includes all return
     * enabled fields (return=_all_fields). To return only the document IDs for
     * the matching documents, specify return=_no_fields. To retrieve the
     * relevance score calculated for each document, specify return=_score. You
     * specify multiple return fields as a comma separated list. For example,
     * return=title,_score returns just the title and relevance score of each
     * matching document.
     *
     * @param  string $returnFields
     * @return StructuredQueryBuilder
     */
    public function returnFields($returnFields)
    {
        $this->returnFields = $returnFields;
        return $this;
    }

    /**
     * SIZE
     * The maximum number of search hits to return.
     *
     * @param  integer $size
     * @return StructuredQueryBuilder
     */
    public function size($size)
    {
        $this->size = $size;
        return $this;
    }

    /**
     * SORT
     * A comma-separated list of fields or custom expressions to use to sort the
     * search results. You must specify the sort direction (asc or desc) for
     * each field. For example, sort=year desc,title asc. You can specify a
     * maximum of 10 fields and expressions. To use a field to sort results, it
     * must be sort enabled in the domain configuration. Array type fields
     * cannot be used for sorting. If no sort parameter is specified, results
     * are sorted by their default relevance scores in descending order:
     * sort=_score desc. You can also sort by document ID (sort=_id) and
     * version (sort=_version).
     *
     * @param  string $field
     * @param  string $direction
     * @return StructuredQueryBuilder
     */
    public function sort($field, $direction = 'asc')
    {
        $this->sort = "{$field} {$direction}";
        return $this;
    }

    /**
     * START
     * The offset of the first search hit you want to return. You can specify
     * either the start or cursor parameter in a request, they are mutually
     * exclusive.
     *
     * @param  integer $start
     * @return StructuredQueryBuilder
     */
    public function start($start)
    {
        $this->start = $start;
        return $this;
    }

    /**
     * STATISTICS
     * To get statistics for a field you use the stats.FIELD parameter. FIELD
     * is the name of a facet-enabled numeric field. You specify an empty JSON
     * object, stats.FIELD={}, to get all of the available statistics for the
     * specified field. (The stats.FIELD parameter does not support any options;
     * you must pass an empty JSON object.) You can request statistics for
     * multiple fields in the same request.
     *
     * You can get statistics only for facet-enabled numeric fields: date,
     * date-array, double, double-array, int, or int-array. Note that only the
     * count, max, min, and missing statistics are returned for date and
     * date-array fields.
     *
     * @param  string $field
     * @return StructuredQueryBuilder
     */
    public function stats($field)
    {
        $this->stats[] = $field;
    }


    /**
     * Special function to filter by distance (lat/lon)
     *
     * @param  string  $field
     * @param  string  $lat
     * @param  string  $lon
     * @param  integer $radius
     * @param  boolean $addExpr
     * @return StructuredQueryBuilder
     */
    public function latlon($field, $lat, $lon, $radius = 50, $addExpr = false)
    {
        // upper left bound
        $lat1 = $lat + ($radius/69);
        $lon1 = $lon - $radius/abs(cos(deg2rad($lat))*69);
        // lower right bound
        $lat2 = $lat - ($radius/69);
        $lon2 = $lon + $radius/abs(cos(deg2rad($lat))*69);

        $min = "'{$lat1},{$lon1}'";
        $max = "'{$lat2},{$lon2}'";
        $this->fq->range($field, $min, $max);
        if ($addExpr) {
            $this->addDistanceExpr($field, $lat, $lon);
        }
        return $this;
    }

    /**
     * Special function to add 'distance' expression
     *
     * @param string $field
     * @param string $lat
     * @param string $lon
     * @return StructuredQueryBuilder
     */
    public function addDistanceExpr($field, $lat, $lon)
    {
        $expression = "haversin(".
          "{$lat},".
          "{$lon},".
          "{$field}.latitude,".
          "{$field}.longitude)";
        $this->expr("distance", $expression);
        return $this;
    }

    /**
     * Build the structured query array to send to AWS search
     *
     * @return array
     */
    public function buildStructuredQuery()
    {
        $structuredQuery;
        // cursor
        if ($this->cursor) {
            $structuredQuery['cursor'] = $this->cursor;
        }
        // expressions
        if ($this->expressions) {
            $structuredQuery['expr'] = json_encode($this->expressions);
        }
        // facets
        if ($this->facets) {
            $structuredQuery['facet'] = json_encode($this->facets);
        }
        // filter query
        if ($this->fq->query) {
            $structuredQuery['filterQuery'] = (string)$this->fq;
        }
        // pretty output
        if (!is_null($this->pretty)) {
            $structuredQuery['pretty'] = $this->pretty;
        }
        // query
        if ($this->q->query) {
            $structuredQuery['query'] = (string)$this->q;
        }
        // options
        if ($this->options) {
            $structuredQuery['queryOptions'] = json_encode($this->options);
        }
        // highlights
        // partial
        // parser
        $structuredQuery['queryParser'] = 'structured';
        // return
        if ($this->returnFields) {
            $structuredQuery['return'] = $this->returnFields;
        }
        // size
        $structuredQuery['size'] = $this->size;
        // sort
        if ($this->sort) {
            $structuredQuery['sort'] = $this->sort;
        }
        // start or cursor
        if ($this->start >= 10000 and is_null($this->cursor)) {
            $this->cursor = $this->buildCursor();
        }
        if (!$this->cursor) {
            $structuredQuery['start'] = $this->start;
        }

        // stats
        if ($this->stats) {
            $stats = [];
            foreach($this->stats as $statField) {
                $stats[] = "\"{$statField}\":{}";
            }
            $stats = "{".implode(',',$stats)."}";
            $structuredQuery['stats'] = $stats;
        }
        return $structuredQuery;
    }

    /**
     * Auto setup cursor for query exceeding max 10000 offset
     *
     * @return string
     */
    private function buildCursor()
    {
        // check if this query has already been cached
        // if found in cache, return cursor
        $hash = $this->toHash();
        if ($cursor = Cache::tags(['cursors'])->get($hash)) {
            return $cursor;
        }
        // not found, so check if there's a cursor for the previous page
        $prevBuilder = clone $this;
        $prevBuilder->start($this->start - $this->size);
        if ($prevCursor = Cache::tags(['cursors'])->get($prevBuilder->toHash())) {
            // found it, but now we'll have to fetch another cursor for the
            // current search request
            $cursor = $this->fetchCursor(
                $this, //original query
                $this->size, //just a page worth remaining
                $this->size, //grab a page worth
                $prevCursor, //the previous cursor
                ($this->start - $this->size) //completed all but a page
            );
            return $cursor;
        }
        // we'll run recursive cursor method to build cursor hashes from the start value
        return $this->fetchCursor($this, $this->start);
    }

    /**
     * Run cloudsearch query to retrieve a cursor value
     *
     * @param  StructuredQueryBuilder  $originalBuilder
     * @param  integer  $remaining
     * @param  integer $size
     * @param  string  $seekCursor
     * @param  integer $completed
     * @return string
     */
    private function fetchCursor(
        $originalBuilder,
        $remaining,
        $size = 10000,
        $seekCursor = 'initial',
        $completed = 0
    )
    {
        // determine what we are seeking
        $seekBuilder = clone $originalBuilder;
        $seekBuilder->start($completed + $size);
        // first check in cache
        if (Cache::tags(['cursors'])->has($seekBuilder->toHash())) {
            $cursor = Cache::tags(['cursors'])->get($seekBuilder->toHash());
        } else {
            // go to cloudsearch
            $cursorQuery = clone $seekBuilder;
            $cursorQuery->returnFields('_no_fields')
                        ->cursor($seekCursor)
                        ->size($size);
            // unset facets
            $cursorQuery->facets = [];
            // get cursor from cloudsearch and set to cache
            // get parent config to pass to new CloudSearchQuery instance
            $cursorResults = (new CloudSearchQuery($this->parent->config, $cursorQuery))->get();
            $cursor = $cursorResults->cursor;
            Cache::tags(['cursors'])->forever($seekBuilder->toHash(), $cursor);
        }
        // determine if we need to get more cursors
        $remainingSize = ($remaining - $size);
        if ($remainingSize > 0) {
            /* go get it */
            return $this->fetchCursor(
                $originalBuilder,
                $remainingSize,
                ($remainingSize < 10000) ? $remainingSize : 10000,
                $cursor,
                ($completed + $size)
            );
        } else {
            /* We arrived, return the cursor */
            return $cursor;
        }
    }

    /**
     * Create unique hash for builder
     *
     * @return string
     */
    private function toHash()
    {
        $hash = hash('md4', json_encode($this));
        return $hash;
    }


}
